### PR TITLE
[MUGEN] Add Text-to-Video GPT Model Builder

### PR DESCRIPTION
**Pull request by langong347**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #282
* #281
* __->__ #276
* #283

Summary:
- Added `text_to_video_gpt` builder for text to video generation
- Added public attributes to `VQVAE` and `MultimodalGPT`
    - `num_embeddings`: codebook size --> needed to create projections that bridge the tokenizers and the transformer
    - `embedding_dim`: size of each codebook vector --> same as above
    - `num_{in/out}_tokens`: number of token states (for both the in- and out- modalities) --> needed in #282 to define the generation states
- Added option to freeze `video_vqvae_mugen` after loading MUGEN checkpoints
    
Test Plan:
- Temporarily modified `.coveragerc` to show reports for `examples/`
```
$ python -m pytest --cov=examples/mugen/generation/ examples/mugen/test/generation/test_text_video_gpt.py examples/mugen/test/generation/test_video_vqvae.py -vv
================================================= test session starts ==================================================
platform darwin -- Python 3.8.13, pytest-7.1.2, pluggy-1.0.0 -- /Users/langong/local/miniconda3/envs/t2v/bin/python
cachedir: .pytest_cache
rootdir: /Users/langong/gpt_attention, configfile: pyproject.toml
plugins: mock-3.8.2, cov-3.0.0
collected 26 items

examples/mugen/test/generation/test_text_video_gpt.py::test_encode_text PASSED                                   [  3%]
examples/mugen/test/generation/test_text_video_gpt.py::test_encode_video[8-expected0] PASSED                     [  7%]
examples/mugen/test/generation/test_text_video_gpt.py::test_encode_video[16-expected1] PASSED                    [ 11%]
examples/mugen/test/generation/test_text_video_gpt.py::test_encode_video[32-expected2] PASSED                    [ 15%]
examples/mugen/test/generation/test_text_video_gpt.py::test_decode_video[8-55462.1641] PASSED                    [ 19%]
examples/mugen/test/generation/test_text_video_gpt.py::test_decode_video[16-112028.1719] PASSED                  [ 23%]
examples/mugen/test/generation/test_text_video_gpt.py::test_decode_video[32-225157.7656] PASSED                  [ 26%]
examples/mugen/test/generation/test_text_video_gpt.py::test_decode_video_checkpoint[8-116013.4766] PASSED        [ 30%]
examples/mugen/test/generation/test_text_video_gpt.py::test_decode_video_checkpoint[16-237488.625] PASSED        [ 34%]
examples/mugen/test/generation/test_text_video_gpt.py::test_decode_video_checkpoint[32-536481.4375] PASSED       [ 38%]
examples/mugen/test/generation/test_text_video_gpt.py::test_lookup[in-expected_shape0--53.7916] PASSED           [ 42%]
examples/mugen/test/generation/test_text_video_gpt.py::test_lookup[out-expected_shape1-42.4742] PASSED           [ 46%]
examples/mugen/test/generation/test_text_video_gpt.py::test_forward[8-80.8206] PASSED                            [ 50%]
examples/mugen/test/generation/test_text_video_gpt.py::test_forward[16-9.1838] PASSED                            [ 53%]
examples/mugen/test/generation/test_text_video_gpt.py::test_forward[32-43.5649] PASSED                           [ 57%]
examples/mugen/test/generation/test_text_video_gpt.py::test_forward_checkpoint[8-80.8206] PASSED                 [ 61%]
examples/mugen/test/generation/test_text_video_gpt.py::test_forward_checkpoint[16-9.1838] PASSED                 [ 65%]
examples/mugen/test/generation/test_text_video_gpt.py::test_forward_checkpoint[32-43.5649] PASSED                [ 69%]
examples/mugen/test/generation/test_video_vqvae.py::TestVideoVQVAEMUGEN::test_invalid_pretrained_model_key PASSED [ 73%]
examples/mugen/test/generation/test_video_vqvae.py::TestVideoVQVAEMUGEN::test_freeze_model PASSED                [ 76%]
examples/mugen/test/generation/test_video_vqvae.py::TestVideoVQVAEMUGEN::test_forward[8--16005.1113] PASSED      [ 80%]
examples/mugen/test/generation/test_video_vqvae.py::TestVideoVQVAEMUGEN::test_forward[16-51112.2266] PASSED      [ 84%]
examples/mugen/test/generation/test_video_vqvae.py::TestVideoVQVAEMUGEN::test_forward[32-450732.6875] PASSED     [ 88%]
examples/mugen/test/generation/test_video_vqvae.py::TestVideoVQVAEMUGEN::test_checkpoint[8-132017.28125] PASSED  [ 92%]
examples/mugen/test/generation/test_video_vqvae.py::TestVideoVQVAEMUGEN::test_checkpoint[16--109636.0] PASSED    [ 96%]
examples/mugen/test/generation/test_video_vqvae.py::TestVideoVQVAEMUGEN::test_checkpoint[32-1193122.0] PASSED    [100%]

---------- coverage: platform darwin, python 3.8.13-final-0 ----------
Name                                          Stmts   Miss  Cover
-----------------------------------------------------------------
examples/mugen/generation/text_video_gpt.py      69      8    88%
examples/mugen/generation/video_vqvae.py         24      0   100%
-----------------------------------------------------------------
TOTAL                                            93      8    91%


================================ 26 passed in 46.01s ======================
```

Differential Revision: [D38865165](https://our.internmc.facebook.com/intern/diff/D38865165)

### CODE REVIEW

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38865165).

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38865165).

File: examples/mugen/generation/text_video_gpt.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        """Decodes token ids back to sentences"""
+        sentences = []
+        for _token_ids in token_ids:  # iterate over batches
+            _token_ids = self._filter_token_ids(_token_ids.tolist())
+            sentence = self.tokenizer.decode(_token_ids)  # type: ignore
+            sentences.append(sentence)
+
+        return sentences
+
+    def lookup(self, token_ids: Tensor) -> Tensor:
```
Comment (ebsmothers) [inline examples/mugen/generation/text_video_gpt.py:224]: Just wondering, why do we include embeddings and lookup in this class? Seems without them we would basically just have a tokenizer class and and embedding module, then we could bundle these together in MultimodalGPT and just use the embedding directly there. (Or is there some case where we need both the tokenizer and the embedding logic in the same class?)

File: examples/mugen/generation/text_video_gpt.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        latent_shape=video_latent_shape,
+        in_tokenizer=text_tokenizer,
+        out_tokenizer=video_vqvae,
+        mm_decoder=mm_decoder,
+        in_projection=text_projection,
+        out_projection=video_projection,
+    )
+
+
+def latent_shape(
```
[the remark this answers is not in the document]
Comment (langong347) [reply examples/mugen/generation/text_video_gpt.py:152]: TL;DR

This is a stop-gap solution to ensure that `latent_shape` inferred by the video encoder (of VQVAE) is consistent with the actual `latent_shape` (i.e., the one derived from the data).

Long Answer:

`latent_shape` is the ratio between the input data (from the dataset) and downsample rates, e.g., an input of `(8, 32, 32)` with downsample rates `(2, 4, 4)` will have latent shape `(4, 8, 8)`.

Ideally, the encoder of `video_vqvae` should be constructed from downsample rates as the number of convolution layers and strides (in each layer) should be consistent with number of times to downsample, i.e., `math.log2(down_sample)` in each dim.

However, our current `video_vqvae` does not take `downsample` rates as an input such  that we can only back out `latent_shape` from the encoder's architecture:

https://github.com/facebookresearch/multimodal/blob/258b5b573d5040d01d5eb54b754a44243c9afaf0/torchmultimodal/models/video_vqvae.py#L164-L173

File: examples/mugen/generation/text_video_gpt.py (hunk 1 above)
Comment (langong347) [reply examples/mugen/generation/text_video_gpt.py:224]: This is a design choice to align the APIs of different tokenizers to be suitable for generation.

`MultimodalGPT` requires the modality-specific tokenizers to implement the same [APIs](https://github.com/facebookresearch/multimodal/blob/258b5b573d5040d01d5eb54b754a44243c9afaf0/torchmultimodal/models/gpt.py#L235-L303):
- `encode`: data to token ids
- `lookup`: token ids to embedding vectors
- `decode`: token ids to data

Although a text tokenizer is strictly for tokenization, it needs to have the above APIs to be consistent with the video tokenizer (VQVAE) where tokenization and embedding (2 in 1) both come from `Codebook`:

https://github.com/facebookresearch/multimodal/blob/258b5b573d5040d01d5eb54b754a44243c9afaf0/torchmultimodal/modules/layers/codebook.py#L261-L263

File: examples/mugen/generation/text_video_gpt.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        return [
+            token_id
+            for token_id in token_ids
+            if token_id > 0 and token_id <= self.vocab_size
+        ]
+
+    def decode(self, token_ids: Tensor) -> List[str]:
+        """Decodes token ids back to sentences"""
+        sentences = []
+        for _token_ids in token_ids:  # iterate over batches
```
[the remark this answers is not in the document]
Comment (langong347) [reply examples/mugen/generation/text_video_gpt.py:217]: offline discussion: within the library we have been assuming batch first. will just follow that convention.

File: examples/mugen/generation/text_video_gpt.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        )
+
+    # builds text embedding projection: text_emb is already of output shape `d_model`
+    # generally a projection layer is needed to bridge the tokenizer and
+    # `torchmultimodal.models.gpt.MultimodalTransformerDecoder`, see `video_projection`
+    text_projection = nn.Identity()
+
+    # builds video embedding projection
+    video_projection = nn.Linear(video_vqvae.embedding_dim, d_model, bias=False)
+    video_projection.weight.data.normal_(std=0.02)
```
[the remark this answers is not in the document]
Comment (langong347) [reply examples/mugen/generation/text_video_gpt.py:124]: will move weight initialization to `MultimodalGPT` following CLIP example.

File: examples/mugen/generation/text_video_gpt.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    def encode(self, sentences: List[str], device: str) -> Tensor:
+        """Encodes sentences to token ids"""
+        token_ids = self.text_to_tokens(sentences).to(device)
+        # bump padding token ids by vocab_size so that they do not conincide with un-padded token ids
+        # and that the padding token ids themselves are unique
+        unique_pad_ids = torch.arange(self.context_len, device=device) + self.vocab_size
+        token_ids = torch.where(token_ids == self.pad_id, unique_pad_ids, token_ids)
+        return token_ids
+
+    def _filter_token_ids(self, token_ids: List[int]) -> List[Optional[int]]:
```
[the remark this answers is not in the document]
Comment (langong347) [reply examples/mugen/generation/text_video_gpt.py:206]: Since the filtered token sequences are of uneven lengths or empty, the return type is `List[Optional[int]]` instead of `List[int]` or `Tensor`.

**Review by ebsmothers — APPROVED** [review]

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38865165).
